### PR TITLE
Feature/ui updates

### PR DESCRIPTION
**Pull request by Gibbo3771**
This is basically the foundation for darkmode across the entire UI. So far I have done darkmode for everything _except_ the icons as they are part of the common library. It would be more appropriate to modify that code to accept a color palette than shim it in here.

I also can't control the bottom bar along the computer terminal display for this same reason, but over all it looks pretty good.

I also forgot the `.java-version` file. This will allow mac users to install `jenv` and set their shell java path to the required version (like rbenv or pipenv).

### CODE REVIEW

**Comment by Gibbo3771** [comment]
Worth noting I had considered transforming the current CSS to SASS but decided against it as I didn't feel it was worth replacing all the css modules.

**Review by SquidDev — CHANGES_REQUESTED** [review]
Thanks for looking into this!

I wonder instead of threading `Settings` everywhere through the app, it'd be cleaner just to toggle a class/attribute on `document.body` and handle everything else through CSS.

The recommended approach on [css tricks](https://css-tricks.com/a-complete-guide-to-dark-mode-on-the-web/) at least seems to be doing that+CSS variables. CSS route also means we can use `@media (prefers-color-scheme: dark)` for extra fanciness.

> So far I have done dark-mode for everything except the icons [and] the bottom bar along the computer terminal display

As I'm sure you've noticed, that'll require changes to `cc-web-term`. If we shift that to using CSS variables for colours, should be pretty easy to get that to respond to theme changes.

> Worth noting I had considered transforming the current CSS to SASS but decided against it as I didn't feel it was worth replacing all the css modules.

It should be pretty easy to get this working with `rollup-plugin-postcss`, but not sure how it'd play with our dodgy generation of `styles.d.ts`. As you say, probably not worth it :p.

> I also forgot the .java-version file. This will allow mac users to install jenv and set their shell java path to the required version (like rbenv or pipenv).

Ooooh, nice! I was somewhat hoping I could do this with Gradle's `toolchain`, but looks like it'll require a bit more work to get it to work with TeaVM.

File: src/web/ts/settings.tsx [hunk 1]
```diff
@@ -153,46 +155,66 @@ export type SettingsProperties = {
...
-  <div class={dialogueBox}>
-    <h2>Settings</h2>
+export const Settings = ({ store, configGroups }: SettingsProperties): JSX.Element | null => {
+  const [darkMode, setDarkMode] = useState<boolean | undefined>(undefined);
+
+  useEffect(() => {
+    for(const g of configGroups){
+      for(const property of g.properties){
+        if(property.id === "editor.dark") {
+          // I have no idea how this works
```
[the remark this answers is not in the document]
Comment (Gibbo3771) [reply src/web/ts/settings.tsx:165]: I would then need to send the `darkMode` prop into `settings.tsx` is the only thing.

**Comment by Gibbo3771** [comment]
> I wonder instead of threading Settings everywhere through the app, it'd be cleaner just to toggle a class/attribute on document.body and handle everything else through CSS.

I was actually thinking about this, I notice there is quite a lot of prop drilldown. Also settings is never truly reactive because it is not part of application state. this is why I have a `darkMode` state in the `Settings` component.

Settings could be extracted into a custom hook that manages the settings state, then you can wrap the entire thing in a provider, but that's quite a bit of work. 

Overall I think having SCSS + a theme provider controlled by `prefers-color-scheme` is the way forward. However, it does not solve the problem that there places where components need to be supplimented, like so:

```jsx
<div class={clsx(dialogueBox, {[dialogueBoxDark]: darkMode})}> 
    <h2 class={clsx({[fontDarkMode]: darkMode})}>Settings</h2>
    {configGroups.map(({ name, description, properties }) => [
      <h3 class={clsx({[fontDarkMode]: darkMode})}>{name}</h3>,
      description ? <p class={clsx(tinyText, {[fontDarkMode]: darkMode})}>{description}</p> : null,
      <div class={clsx(formGroup, {[formGroupDark]: darkMode})}>
        {properties.map(property => {
          switch (property.type) {
            case "string":
              return <label>
                {property.name}
                <input type="text" value={store.get(property)}
                  onChange={getUpdater(store, property, getString)}></input>
                <p class={clsx(tinyText, {[tinyTextDark]: darkMode})}>{property.description}</p>
              </label>;
            case "int":
              return <label>
                {property.name}
                <input type="number" value={store.get(property)} min={property.min} max={property.max} step={1}
                  onChange={getUpdater(store, property, getNumber)}></input>
                <p class={clsx(tinyText, {[tinyTextDark]: darkMode})}>{property.description}</p>
              </label>;
            case "boolean":
              return <label>
                <input type="checkbox" checked={store.get(property)}
                  onInput={getUpdater(store, property, getBool)}></input>
                {property.name}
                <p class={clsx(tinyText, {[tinyTextDark]: darkMode})}>{property.description}</p>
              </label>;
            case "option":
              return <label>
                {property.name}
                <select value={store.get(property)} onInput={getUpdater(store, property, getOption(property.def, property.choices))}>
                  {property.choices.map(({ key, value }) => <option value={key}>{value}</option>)}
                </select>
                <p class={clsx(tinyText, {[tinyTextDark]: darkMode})}>{property.description}</p>
              </label>;
          }
        })}
      </div>,
    ])}
  </div>}
```

Although we can modify `p`, `h2` etc directly, the "wrapper" tag will need to be extracted out into a component that knows how to render the theme.

File: src/web/ts/settings.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (Gibbo3771) [reply src/web/ts/settings.tsx:165]: Disregard, I see what you mean now

**Comment by SquidDev** [comment]
> Also settings is never truly reactive[...] Settings could be extracted into a custom hook that manages the settings state, then you can wrap the entire thing in a provider, but that's quite a bit of work.

Not going to pretend this any of my projects follow React/Preact best practices. Wait 'til you see the horror which is the terminal renderer :p.

That said, also not a massive fan of hooks, so if we can avoid them all the better. As you say, probably not worth it for now.

>  However, it does not solve the problem that there places where components need to be supplemented, like so:

I'm possibly a little lost here, but not sure what additional stuff in this example is needed. Assumedly we can drop the `*Dark` classes and just replace with something like:


```css
body { --tiny-text-colour: #555; }
body.dark-mode { --tiny-text-colour: rgb(231, 231, 231); } /* And/or @media */
.tinyText { color: var(--tiny-text-colour); }
```

<sub>But with decent formatting</sub>

**Comment by Gibbo3771** [comment]
I have updated this to use CSS class vars to see what you think. I tried to split the main colors from the specific bits of css as you can see. Unfortunately without SASS, dealing with variables and classes is just messy.

**Comment by Gibbo3771** [comment]
> I'm possibly a little lost here, but not sure what additional stuff in this example is needed.

Yeah that's fine, I was talking out my ass. All my React projects are either written using css-in-js or sass, so I take advantage of state based themes. I usually use a theme context and provider which allows me to write hooks that components can use to setup their css-in-js. This allows a set color palette across the board but then allows you to do other stuff, such as user created themes and multiple layouts and what not.

Basically, ignore my comment.

**Review by SquidDev — APPROVED** [review]

**Pull request merged**